### DISCUSSION TITLE
Unable to load Jaro-Winkler Similarity in AWS Glue

### DISCUSSION THREAD

**Discussion by richard-a-lott** [moj-analytical-services/splink#1377] [Q&A]
Hi,

I've had a bit of an issue using Splink with AWS Glue (similar to this one: https://github.com/moj-analytical-services/splink/discussions/636#discussion-4211560) where Spark is unable to find and load the Splink UDF jar.

I confirmed the jar was present in the path identified by similarity_jar_location, and I confirmed the jar was included in multiple spark config parameters (spark.jars, spark.driver.extraCLassPath, spark.executor.extraClassPath), however my Glue job refused to recognise the jar.

I have found a work-around, which I'll share here, but it would be good to know if anyone has found a better way?

My work around is to specify, in the Glue Job Details, a location for "Dependend Jars Path" (has to be full S3 URI to the jar file, including the filename), and then to upload the jar file into this S3 location, at the start of the Glue job, prior to creating the Spark session:

![image](https://github.com/moj-analytical-services/splink/assets/136445689/6d13f437-b75e-4931-8e08-1a1adda0b039)

![image](https://github.com/moj-analytical-services/splink/assets/136445689/28866663-c236-4d91-8d6b-956c82932e36)
(apologies for the image, I'm unable to paste the code in directly)

Maybe a bit messy but it seems to be working. Anyone have a better method?

**Comment by richard-a-lott** [comment]
Actually this _doesn't (quite)_ work! One has to copy the jar file into S3 _before_ running the job. The rest of the method works though.

In my case I'd already copied the jar, so it existed before the job ran.

So a revised method to the above is:
- Copy the jar file to a target S3 location - manually or via another process prior to your Glue job
- Reference the S3 location in the Glue job parameter '--extra-jars'

**Comment by 93Akkord** [comment]
Hmmm.. I've been able to use and load the jar file just fine.  Are you including splink into your job in the job parameters section like below?

![image](https://github.com/moj-analytical-services/splink/assets/8103207/c9945c27-4f30-4396-b41a-198c6929f501)

**Reply by richard-a-lott** [reply]
Yes I'm doing this (version 3.9.2 though), and the jar is definitely present in the Glue environment, but spark wasn't loading it, even after specifying it as per the example in [here](https://moj-analytical-services.github.io/splink/demos/example_simple_pyspark.html).

Let me check again in case I've done something stupid.

**Comment by richard-a-lott** [comment]
No it's still not working. I can't see what I've done wrong. Here's an overview of my spark config:
![image](https://github.com/moj-analytical-services/splink/assets/136445689/4e9309a3-d3a6-4ecc-a3ee-e7dfea9710ca)

Here is the log showing it's being picked up by spark:
[Edit: Image removed]

And here's the susequent log showing it's failed to load the udf:
[Edit: Image removed]

I'm baffled. Any ideas?

**Reply by RobinL** [reply]
Hello.  I've just been looking at how we do this.  Sorry, this will be a little confusing because our methodology is a legacy methodology that we started using before Splink 3.  

I'll try and describe (1) our current working methodology and (2) what we should be doing if we started afresh.


## Current working methodology (legacy, not best practice)

We use our `etl_manager` package see [here](https://github.com/moj-analytical-services/etl_manager#using-the-gluejob-class) to run our glue jobs programatically using the boto API.  

One of the main purposes of `etl_manager` is to eliminate the anoying work of having to sync your job, your jars, and any additional python files in s3.  Here's a typical API call to boto:

```
{   'AllocatedCapacity': 5,
    'Command': {   'Name': 'glueetl',
                   'PythonVersion': '3',
                   'ScriptLocation': 's3://long_path/job.py'},
    'DefaultArguments': {   '--TempDir': 's3://long_path/glue_temp_folder/',
                            '--additional-python-modules": "dataengineeringutils3==1.4.0,watchtower==3.0.0,splink==3.3.8'
                            '--extra-jars': 's3://long_path/scala-udf-similarity-0.0.8.jar',
                            '--extra-py-files': 's3://long_path/generic_standardisation_funcs.py,s3://long_path/misc_shared_fns.py',
                            '--job-bookmark-option': 'job-bookmark-disable'},
    'ExecutionProperty': {'MaxConcurrentRuns': 1},
    'GlueVersion': '3.0',
    'MaxRetries': 0,
    'Name': 'model_training_person_mags_hocas_no_companies_00_estimate_model',
    'Role': 'your-role-here',
    'Tags': {},
    'Timeout': 545}
```
As you can see, we specify `'--extra-jars': 's3://long_path/scala-udf-similarity-0.0.8.jar',` which puts the udf jar on the spark.jars path automatically, with no need for a call to `conf.set('spark.jars', x)`

Then, in the job, we can run the following function to register all the udfs in the jar, which we run in all our jobs 'just in case' they're needed:

<details>
<summary>Add udfs function</summary>

```python
from pyspark.sql.types import DoubleType, StringType, StructField, ArrayType, StructType


def add_udfs(spark):
    # Register UDFs
    udfs = [
        ("jaro_winkler_sim", "JaroWinklerSimilarity", DoubleType()),
        ("jaccard_sim", "JaccardSimilarity", DoubleType()),
        ("cosine_distance", "CosineDistance", DoubleType()),
        ("Dmetaphone", "DoubleMetaphone", StringType()),
        ("Dmetaphone", "DoubleMetaphone", StringType()),
        ("DmetaphoneAlt", "DoubleMetaphoneAlt", StringType()),
        ("QgramTokeniser", "QgramTokeniser", StringType()),
    ]

    for a, b, c in udfs:
        spark.udf.registerJavaFunction(a, "uk.gov.moj.dash.linkage." + b, c)

    # Register UDFs
    rt = ArrayType(
        StructType([StructField("_1", StringType()), StructField("_2", StringType())])
    )

    spark.udf.registerJavaFunction(
        name="DualArrayExplode",
        javaClassName="uk.gov.moj.dash.linkage.DualArrayExplode",
        returnType=rt,
    )

    rt2 = ArrayType(
        StructType(
            (
                StructField(
                    "place1",
                    StructType(
                        (
                            StructField("lat", DoubleType()),
                            StructField("long", DoubleType()),
                        )
                    ),
                ),
                StructField(
                    "place2",
                    StructType(
                        (
                            StructField("lat", DoubleType()),
                            StructField("long", DoubleType()),
                        )
                    ),
                ),
            )
        )
    )

    spark.udf.registerJavaFunction(
        name="latlongexplode",
        javaClassName="uk.gov.moj.dash.linkage.latlongexplode",
        returnType=rt2,
    )
```

</details>

Then all the udfs work as expected.

The drawback of this approach is that (implicitly in what `etl_manager` does, you have to manually provide `'--extra-jars': 's3://long_path/scala-udf-similarity-0.0.8.jar',` and ensure it's in S3, and also it won't automatically update when Splink updates the jar that's bundled lin the library)
[I'm still writing this, just don't want to lose what i've already written]

## Best practice

I believe the method you're attempting should work with a bit of fiddling.  i.e. 

Specifically, as metioned by @93Akkord you can `pip install splink` using `--additional-python-modules`, and use `similarity_jar_location()` to get the path to the jar

The only caveat here is that, since we don't do this ourselves, I'm not 100% it works (but I believe it does). In particular, the jar needs to be available on every machine in the cluster, so I'm _assuming_ the pip install achieves this.

The specific syntax you need (if you're using Spark 3, which you will be if you're using the [latest version of Glue](https://docs.aws.amazon.com/glue/latest/dg/release-notes.html)) is:

```
conf.set('spark.jars', similarity_jar_location())
```

and setting 

```
--additional-python-modules='splink==3.9.2'"
```
i.e. note that in Spark 3 you shouldn't need to set `extraClassPath` I don't think.

Then note that `Splink` [registers the udfs for you](https://github.com/moj-analytical-services/splink/blob/56833b6fe6692de72530083f51dfdbad29c0fd33/splink/spark/linker.py#L260) you instantiate the linker and will show an [warning](https://github.com/moj-analytical-services/splink/blob/56833b6fe6692de72530083f51dfdbad29c0fd33/splink/spark/linker.py#L289) if this does not work.  i.e. you shouldn't need to do this yourself.

❗️ This also means that if you must create the `linker = SparkLinker()` _after_ running `conf.set('spark.jars', similarity_jar_location())`

@93Akkord Thanks for your earlier response.  Does the above 'best practice' methodology sound right to you?  i.e. is this the methodology you're using to get it to work?

**Reply by richard-a-lott** [reply]
Thanks Robin, your best practice is what I was originally trying. I spent most of yesterday experimenting with this approach, plus a variety of different spark config parameters and getting nowhere.

I can see that the jar is present in the Glue workspace, and I can see the log showing that the jar is being loaded by Spark, but splink is still giving the error. The only way I can find of making it work is to copy the jar ahead of time to S3, and specify it in the job dependent jars parameter.

This _could_ be down to the AWS environment I'm using. It's particularly locked down and I have very little access to anything. I note that the log is referencing an IP address and i'm thinking perhaps I don't have access:
![image](https://github.com/moj-analytical-services/splink/assets/136445689/9dbd9511-2baa-4765-9872-96a9a632610e)

It's only a guess, though

**Reply by RobinL** [reply]
Do you have access to AWS support via your organisation (i.e. raising a ticket on the AWS website)?  We've had a lot of success before raising tickets for advice - their engineers are usually prompt in replying and have a lot of expertise (even just for help with something like this, rather than a bug).  

If you don't, we should be able to raise a ticket at our end, because we also would like to know whether there's a way of getting the 'best practice' suggestion to work.

In particular, they should be able to confirm/refute whether the `pip install` route makes the jar available on all machines (drivers and worker), and hopefully explain the source of the confusing error you're seeing.

**Reply by richard-a-lott** [reply]
My first line of support is the team that manages our AWS environment, so I will raise a question with them, however whether they are able to answer (and willing to make any required access changes) is another question!

If you wanted to raise this separately with AWS then you might get a more comprehensive answer. I'm happy to provide what info I can, if you want to go down this route.

**Reply by richard-a-lott** [reply]
@RobinL I've been following up this issue with AWS support (via my internal support team) and after a bit of investigation they have come back with the following response. It seems that the 'best practice' solution suggested above should not work, and the jar will need to be in S3 before running, in order for the job to work properly.

I'm happy with my current working method. I just wanted to share this with you for info.

AWS response:
> [...]
> The team confirmed that Glue does not make use of the spark.jars parameter. The Glue job builds a separate classpath used to start the individual executors before connecting to the driver. Once started, this classpath cannot be changed. In other words, even though we can access the .jar file from the script on the driver, changing the classpath already on the executors themselves to point to the jar is not supported.
> 
> In order to make sure the latest jar is always on the UDF, it is instead recommended to regularly update the jar in S3 to the latest/desired version; replacing the jar with a file of the same name would allow the Glue job to still use the latest jar without making changes to the job itself. Alternatively, the path in the ‘--extra-jars’ parameter can be updated to point to the latest desired version of the jar in S3[1]. 
> 
> References:
> ============
> [1] https://docs.aws.amazon.com/glue/latest/dg/aws-glue-programming-etl-glue-arguments.html

**Reply by RobinL** [reply]
Thanks - that's super useful and seems to explain the annoying behaviour we've been observing!

 So basically means there's no way for us to write aws glue compatible code that auto registers the jar

@samnlindsay worth you being aware

**Reply by richard-a-lott** [reply]
The AWS response states that it's not possible to alter the classpath. If there were a way to copy the jar into a location already on the classpath that might work (but I don't think it's so easy to move files around on the executors is it?)

**Reply by RobinL** [reply]
That's a good idea, hadn't thought of that.  Worth a try
